File: test/qdmi/test_fomac.cpp
/*
 * Copyright (c) 2023 - 2025 Chair for Design Automation, TUM
 * Copyright (c) 2025 Munich Quantum Software Company GmbH
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

#include "qdmi/FoMaC.hpp"

#include <algorithm>
#include <gtest/gtest.h>
#include <new>
#include <qdmi/client.h>
#include <stdexcept>
#include <string>
#include <tuple>
#include <vector>

namespace fomac {
class DeviceTest : public testing::TestWithParam<Device> {
protected:
  Device device;

  DeviceTest() : device(GetParam()) {}
};

class SiteTest : public DeviceTest {
protected:
  Site site;

  SiteTest() : site(device.getSites().front()) {}
};

class OperationTest : public DeviceTest {
protected:
  Operation operation;

  OperationTest() : operation(device.getOperations().front()) {}
};

TEST(FoMaCTest, StatusToString) {
  EXPECT_EQ(toString(QDMI_WARN_GENERAL), "General warning");
  EXPECT_EQ(toString(QDMI_SUCCESS), "Success");
  EXPECT_EQ(toString(QDMI_ERROR_FATAL), "A fatal error");
  EXPECT_EQ(toString(QDMI_ERROR_OUTOFMEM), "Out of memory");
  EXPECT_EQ(toString(QDMI_ERROR_NOTIMPLEMENTED), "Not implemented");
  EXPECT_EQ(toString(QDMI_ERROR_LIBNOTFOUND), "Library not found");
  EXPECT_EQ(toString(QDMI_ERROR_NOTFOUND), "Element not found");
  EXPECT_EQ(toString(QDMI_ERROR_OUTOFRANGE), "Out of range");
  EXPECT_EQ(toString(QDMI_ERROR_INVALIDARGUMENT), "Invalid argument");
  EXPECT_EQ(toString(QDMI_ERROR_PERMISSIONDENIED), "Permission denied");
  EXPECT_EQ(toString(QDMI_ERROR_NOTSUPPORTED), "Not supported");
  EXPECT_EQ(toString(QDMI_ERROR_BADSTATE), "Bad state");
  EXPECT_EQ(toString(QDMI_ERROR_TIMEOUT), "Timeout");
  // NOLINTNEXTLINE(clang-analyzer-optin.core.EnumCastOutOfRange)
  EXPECT_EQ(toString(static_cast<QDMI_STATUS>(-999)), "Unknown status code");
}

TEST(FoMaCTest, SitePropertyToString) {
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_INDEX), "QDMI_SITE_PROPERTY_INDEX");
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_T1), "QDMI_SITE_PROPERTY_T1");
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_T2), "QDMI_SITE_PROPERTY_T2");
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_NAME), "QDMI_SITE_PROPERTY_NAME");
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_XCOORDINATE),
            "QDMI_SITE_PROPERTY_XCOORDINATE");
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_YCOORDINATE),
            "QDMI_SITE_PROPERTY_YCOORDINATE");
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_ZCOORDINATE),
            "QDMI_SITE_PROPERTY_ZCOORDINATE");
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_ISZONE), "QDMI_SITE_PROPERTY_ISZONE");
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_XEXTENT), "QDMI_SITE_PROPERTY_XEXTENT");
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_YEXTENT), "QDMI_SITE_PROPERTY_YEXTENT");
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_ZEXTENT), "QDMI_SITE_PROPERTY_ZEXTENT");
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_MODULEINDEX),
            "QDMI_SITE_PROPERTY_MODULEINDEX");
  EXPECT_EQ(toString(QDMI_SITE_PROPERTY_SUBMODULEINDEX),
            "QDMI_SITE_PROPERTY_SUBMODULEINDEX");
  // NOLINTNEXTLINE(clang-analyzer-optin.core.EnumCastOutOfRange)
  EXPECT_EQ(toString(static_cast<QDMI_Site_Property>(-999)),
            "QDMI_SITE_PROPERTY_UNKNOWN");
}

TEST(FoMaCTest, OperationPropertyToString) {
  EXPECT_EQ(toString(QDMI_OPERATION_PROPERTY_NAME),
            "QDMI_OPERATION_PROPERTY_NAME");
  EXPECT_EQ(toString(QDMI_OPERATION_PROPERTY_QUBITSNUM),
            "QDMI_OPERATION_PROPERTY_QUBITSNUM");
  EXPECT_EQ(toString(QDMI_OPERATION_PROPERTY_PARAMETERSNUM),
            "QDMI_OPERATION_PROPERTY_PARAMETERSNUM");
  EXPECT_EQ(toString(QDMI_OPERATION_PROPERTY_DURATION),
            "QDMI_OPERATION_PROPERTY_DURATION");
  EXPECT_EQ(toString(QDMI_OPERATION_PROPERTY_FIDELITY),
            "QDMI_OPERATION_PROPERTY_FIDELITY");
  EXPECT_EQ(toString(QDMI_OPERATION_PROPERTY_INTERACTIONRADIUS),
            "QDMI_OPERATION_PROPERTY_INTERACTIONRADIUS");
  EXPECT_EQ(toString(QDMI_OPERATION_PROPERTY_BLOCKINGRADIUS),
            "QDMI_OPERATION_PROPERTY_BLOCKINGRADIUS");
  // NOLINTNEXTLINE(clang-analyzer-optin.core.EnumCastOutOfRange)
  EXPECT_EQ(toString(static_cast<QDMI_Operation_Property>(-999)),
            "QDMI_OPERATION_PROPERTY_UNKNOWN");
}

TEST(FoMaCTest, DevicePropertyToString) {
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_NAME), "QDMI_DEVICE_PROPERTY_NAME");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_VERSION),
            "QDMI_DEVICE_PROPERTY_VERSION");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_STATUS),
            "QDMI_DEVICE_PROPERTY_STATUS");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_LIBRARYVERSION),
            "QDMI_DEVICE_PROPERTY_LIBRARYVERSION");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_QUBITSNUM),
            "QDMI_DEVICE_PROPERTY_QUBITSNUM");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_SITES), "QDMI_DEVICE_PROPERTY_SITES");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_OPERATIONS),
            "QDMI_DEVICE_PROPERTY_OPERATIONS");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_COUPLINGMAP),
            "QDMI_DEVICE_PROPERTY_COUPLINGMAP");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_NEEDSCALIBRATION),
            "QDMI_DEVICE_PROPERTY_NEEDSCALIBRATION");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_LENGTHUNIT),
            "QDMI_DEVICE_PROPERTY_LENGTHUNIT");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_LENGTHSCALEFACTOR),
            "QDMI_DEVICE_PROPERTY_LENGTHSCALEFACTOR");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_DURATIONUNIT),
            "QDMI_DEVICE_PROPERTY_DURATIONUNIT");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_DURATIONSCALEFACTOR),
            "QDMI_DEVICE_PROPERTY_DURATIONSCALEFACTOR");
  EXPECT_EQ(toString(QDMI_DEVICE_PROPERTY_MINATOMDISTANCE),
            "QDMI_DEVICE_PROPERTY_MINATOMDISTANCE");
  // NOLINTNEXTLINE(clang-analyzer-optin.core.EnumCastOutOfRange)
  EXPECT_EQ(toString(static_cast<QDMI_Device_Property>(-999)),
            "QDMI_DEVICE_PROPERTY_UNKNOWN");
}

TEST(FoMaCTest, SessionPropertyToString) {
  EXPECT_EQ(toString(QDMI_SESSION_PROPERTY_DEVICES),
            "QDMI_SESSION_PROPERTY_DEVICES");
  // NOLINTNEXTLINE(clang-analyzer-optin.core.EnumCastOutOfRange)
  EXPECT_EQ(toString(static_cast<QDMI_Session_Property>(-999)),
            "QDMI_SESSION_PROPERTY_UNKNOWN");
}

TEST(FoMaCTest, ThrowIfError) {
  EXPECT_NO_THROW(throwIfError(QDMI_SUCCESS, "Test"));
  EXPECT_NO_THROW(throwIfError(QDMI_WARN_GENERAL, "Test"));
  EXPECT_THROW(throwIfError(QDMI_ERROR_FATAL, "Test"), std::runtime_error);
  EXPECT_THROW(throwIfError(QDMI_ERROR_OUTOFMEM, "Test"), std::bad_alloc);
  EXPECT_THROW(throwIfError(QDMI_ERROR_NOTIMPLEMENTED, "Test"),
               std::runtime_error);
  EXPECT_THROW(throwIfError(QDMI_ERROR_LIBNOTFOUND, "Test"),
               std::runtime_error);
  EXPECT_THROW(throwIfError(QDMI_ERROR_NOTFOUND, "Test"), std::runtime_error);
  EXPECT_THROW(throwIfError(QDMI_ERROR_OUTOFRANGE, "Test"), std::out_of_range);
  EXPECT_THROW(throwIfError(QDMI_ERROR_INVALIDARGUMENT, "Test"),
               std::invalid_argument);
  EXPECT_THROW(throwIfError(QDMI_ERROR_PERMISSIONDENIED, "Test"),
               std::runtime_error);
  EXPECT_THROW(throwIfError(QDMI_ERROR_NOTSUPPORTED, "Test"),
               std::runtime_error);
  EXPECT_THROW(throwIfError(QDMI_ERROR_BADSTATE, "Test"), std::runtime_error);
  EXPECT_THROW(throwIfError(QDMI_ERROR_TIMEOUT, "Test"), std::runtime_error);
  EXPECT_THROW(throwIfError(-999, "Test"), std::runtime_error);
}

TEST_P(DeviceTest, Name) {
  EXPECT_NO_THROW(EXPECT_FALSE(device.getName().empty()));
}

TEST_P(DeviceTest, Version) {
  EXPECT_NO_THROW(EXPECT_FALSE(device.getVersion().empty()));
}

TEST_P(DeviceTest, Status) {
  EXPECT_NO_THROW(std::ignore = device.getStatus());
}

TEST_P(DeviceTest, LibraryVersion) {
  EXPECT_NO_THROW(EXPECT_FALSE(device.getLibraryVersion().empty()));
}

TEST_P(DeviceTest, QubitsNum) {
  EXPECT_NO_THROW(EXPECT_GT(device.getQubitsNum(), 0));
}

TEST_P(DeviceTest, Sites) {
  EXPECT_NO_THROW(EXPECT_FALSE(device.getSites().empty()));
}

TEST_P(DeviceTest, Operations) {
  EXPECT_NO_THROW(EXPECT_FALSE(device.getOperations().empty()));
}

TEST_P(DeviceTest, CouplingMap) {
  EXPECT_NO_THROW(std::ignore = device.getCouplingMap());
}

TEST_P(DeviceTest, NeedsCalibration) {
  EXPECT_NO_THROW(std::ignore = device.getNeedsCalibration());
}

TEST_P(DeviceTest, LengthUnit) {
  EXPECT_NO_THROW(std::ignore = device.getLengthUnit());
}

TEST_P(DeviceTest, LengthScaleFactor) {
  EXPECT_NO_THROW(std::ignore = device.getLengthScaleFactor());
}

TEST_P(DeviceTest, DurationUnit) {
  EXPECT_NO_THROW(std::ignore = device.getDurationUnit());
}

TEST_P(DeviceTest, DurationScaleFactor) {
  EXPECT_NO_THROW(std::ignore = device.getDurationScaleFactor());
}

TEST_P(DeviceTest, MinAtomDistance) {
  EXPECT_NO_THROW(std::ignore = device.getMinAtomDistance());
}

TEST_P(SiteTest, Index) { EXPECT_NO_THROW(std::ignore = site.getIndex()); }

TEST_P(SiteTest, T1) { EXPECT_NO_THROW(std::ignore = site.getT1()); }

TEST_P(SiteTest, T2) { EXPECT_NO_THROW(std::ignore = site.getT2()); }

TEST_P(SiteTest, Name) { EXPECT_NO_THROW(std::ignore = site.getName()); }

TEST_P(SiteTest, XCoordinate) {
  EXPECT_NO_THROW(std::ignore = site.getXCoordinate());
}

TEST_P(SiteTest, YCoordinate) {
  EXPECT_NO_THROW(std::ignore = site.getYCoordinate());
}

TEST_P(SiteTest, ZCoordinate) {
  EXPECT_NO_THROW(std::ignore = site.getZCoordinate());
}

TEST_P(SiteTest, IsZone) { EXPECT_NO_THROW(std::ignore = site.isZone()); }

TEST_P(SiteTest, XExtent) { EXPECT_NO_THROW(std::ignore = site.getXExtent()); }

TEST_P(SiteTest, YExtent) { EXPECT_NO_THROW(std::ignore = site.getYExtent()); }

TEST_P(SiteTest, ZExtent) { EXPECT_NO_THROW(std::ignore = site.getZExtent()); }

TEST_P(SiteTest, ModuleIndex) {
  EXPECT_NO_THROW(std::ignore = site.getModuleIndex());
}

TEST_P(SiteTest, SubmoduleIndex) {
  EXPECT_NO_THROW(std::ignore = site.getSubmoduleIndex());
}

TEST_P(OperationTest, Name) {
  EXPECT_NO_THROW(EXPECT_FALSE(operation.getName().empty()););
}

TEST_P(OperationTest, QubitsNum) {
  EXPECT_NO_THROW(std::ignore = operation.getQubitsNum());
}

TEST_P(OperationTest, ParametersNum) {
  EXPECT_NO_THROW(std::ignore = operation.getParametersNum());
}

TEST_P(OperationTest, Duration) {
  EXPECT_NO_THROW(std::ignore = operation.getDuration());
}

TEST_P(OperationTest, Fidelity) {
  EXPECT_NO_THROW(std::ignore = operation.getFidelity());
}

TEST_P(OperationTest, InteractionRadius) {
  EXPECT_NO_THROW(std::ignore = operation.getInteractionRadius());
}

TEST_P(OperationTest, BlockingRadius) {
  EXPECT_NO_THROW(std::ignore = operation.getBlockingRadius());
}

TEST_P(OperationTest, IdlingFidelity) {
  EXPECT_NO_THROW(std::ignore = operation.getIdlingFidelity());
}

TEST_P(OperationTest, oned) {
  EXPECT_NO_THROW(std::ignore = operation.isZoned());
}

TEST_P(OperationTest, Sites) {
  EXPECT_NO_THROW(std::ignore = operation.getSites());
}

TEST_P(OperationTest, MeanShuttlingSpeed) {
  EXPECT_NO_THROW(std::ignore = operation.getMeanShuttlingSpeed());
}

INSTANTIATE_TEST_SUITE_P(
    // Custom instantiation name
    DeviceTest,
    // Test suite name
    DeviceTest,
    // Parameters to test with
    testing::ValuesIn(FoMaC::getDevices()),
    [](const testing::TestParamInfo<Device>& paramInfo) {
      auto name = paramInfo.param.getName();
      // Replace spaces with underscores for valid test names
      std::ranges::replace(name, ' ', '_');
      return name;
    });

INSTANTIATE_TEST_SUITE_P(
    // Custom instantiation name
    SiteTest,
    // Test suite name
    SiteTest,
    // Parameters to test with
    testing::ValuesIn(FoMaC::getDevices()),
    [](const testing::TestParamInfo<Device>& paramInfo) {
      auto name = paramInfo.param.getName();
      // Replace spaces with underscores for valid test names
      std::ranges::replace(name, ' ', '_');
      return name;
    });

INSTANTIATE_TEST_SUITE_P(
    // Custom instantiation name
    OperationTest,
    // Test suite name
    OperationTest,
    // Parameters to test with
    testing::ValuesIn(FoMaC::getDevices()),
    [](const testing::TestParamInfo<Device>& paramInfo) {
      auto name = paramInfo.param.getName();
      // Replace spaces with underscores for valid test names
      std::ranges::replace(name, ' ', '_');
      return name;
    });
} // namespace fomac

File: test/python/qdmi/test_fomac.py
# Copyright (c) 2023 - 2025 Chair for Design Automation, TUM
# Copyright (c) 2025 Munich Quantum Software Company GmbH
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

"""Test the quantum computation IR."""

from __future__ import annotations

from typing import cast

import pytest

from mqt.core.qdmi.fomac import Device, DeviceStatus, Operation, Site, devices


@pytest.fixture(params=devices())
def device(request: pytest.FixtureRequest) -> Device:
    """Fixture to provide a device for testing.

    Returns:
        Device: A quantum device instance.
    """
    return cast("Device", request.param)


@pytest.fixture(params=devices())
def device_and_site(request: pytest.FixtureRequest) -> tuple[Device, Site]:
    """Fixture to provide a device for testing.

    Returns:
        tuple[Device, Site]: A tuple containing a quantum device instance and one of its sites.
    """
    device = request.param
    site = device.sites()[0]
    return device, site


@pytest.fixture(params=devices())
def device_and_operation(request: pytest.FixtureRequest) -> tuple[Device, Operation]:
    """Fixture to provide a device for testing.

    Returns:
        tuple[Device, Operation]: A tuple containing a quantum device instance and one of its operations.
    """
    device = request.param
    operation = device.operations()[0]
    return device, operation


def test_device_name(device: Device) -> None:
    """Test that the device name is a non-empty string."""
    name = device.name()
    assert isinstance(name, str)
    assert len(name) > 0


def test_device_version(device: Device) -> None:
    """Test that the device version is a non-empty string."""
    version = device.version()
    assert isinstance(version, str)
    assert len(version) > 0


def test_device_status(device: Device) -> None:
    """Test that the device status is a valid DeviceStatus enum member."""
    status = device.status()
    assert isinstance(status, DeviceStatus)


def test_device_library_version(device: Device) -> None:
    """Test that the device library version is a non-empty string."""
    lib_version = device.library_version()
    assert isinstance(lib_version, str)
    assert len(lib_version) > 0


def test_device_qubits_num(device: Device) -> None:
    """Test that the device qubits number is a positive integer."""
    qubits_num = device.qubits_num()
    assert isinstance(qubits_num, int)
    assert qubits_num > 0


def test_device_sites(device: Device) -> None:
    """Test that the device sites is a non-empty list of Site objects."""
    sites = device.sites()
    assert isinstance(sites, list)
    assert len(sites) > 0
    assert all(isinstance(site, Site) for site in sites)


def test_device_operations(device: Device) -> None:
    """Test that the device operations is a non-empty list of Operation objects."""
    operations = device.operations()
    assert isinstance(operations, list)
    assert len(operations) > 0
    assert all(isinstance(op, Operation) for op in operations)


def test_device_coupling_map(device: Device) -> None:
    """Test that the device coupling map is a list of tuples of Site objects."""
    cm = device.coupling_map()
    if cm is not None:
        assert isinstance(cm, list)
        assert all(len(pair) == 2 for pair in cm)
        assert all(isinstance(site, Site) for pair in cm for site in pair)


def test_device_needs_calibration(device: Device) -> None:
    """Test that the device needs calibration is an integer."""
    needs_cal = device.needs_calibration()
    if needs_cal is not None:
        assert isinstance(needs_cal, int)


def test_device_length_unit(device: Device) -> None:
    """Test that the device length unit is a non-empty string."""
    lu = device.length_unit()
    if lu is not None:
        assert isinstance(lu, str)
        assert len(lu) > 0


def test_device_length_scale_factor(device: Device) -> None:
    """Test that the device length scale factor is a positive float."""
    lsf = device.length_scale_factor()
    if lsf is not None:
        assert isinstance(lsf, float)
        assert lsf > 0.0


def test_device_duration_unit(device: Device) -> None:
    """Test that the device duration unit is a non-empty string."""
    du = device.duration_unit()
    if du is not None:
        assert isinstance(du, str)
        assert len(du) > 0


def test_device_duration_scale_factor(device: Device) -> None:
    """Test that the device duration scale factor is a positive float."""
    dsf = device.duration_scale_factor()
    if dsf is not None:
        assert isinstance(dsf, float)
        assert dsf > 0.0


def test_device_min_atom_distance(device: Device) -> None:
    """Test that the device minimum atom distance is a positive float."""
    mad = device.min_atom_distance()
    if mad is not None:
        assert isinstance(mad, int)
        assert mad > 0.0


def test_site_index(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site index is a non-negative integer."""
    _device, site = device_and_site
    index = site.index()
    assert isinstance(index, int)
    assert index >= 0


def test_site_t1(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site T1 coherence time is a positive integer."""
    _device, site = device_and_site
    t1 = site.t1()
    if t1 is not None:
        assert isinstance(t1, int)
        assert t1 > 0


def test_site_t2(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site T2 coherence time is a positive integer."""
    _device, site = device_and_site
    t2 = site.t2()
    if t2 is not None:
        assert isinstance(t2, int)
        assert t2 > 0


def test_site_name(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site name is a non-empty string."""
    _device, site = device_and_site
    name = site.name()
    if name is not None:
        assert isinstance(name, str)
        assert len(name) > 0


def test_site_x_coordinate(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site x coordinate is an integer."""
    _device, site = device_and_site
    x = site.x_coordinate()
    if x is not None:
        assert isinstance(x, int)


def test_site_y_coordinate(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site y coordinate is an integer."""
    _device, site = device_and_site
    y = site.y_coordinate()
    if y is not None:
        assert isinstance(y, int)


def test_site_z_coordinate(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site z coordinate is an integer."""
    _device, site = device_and_site
    z = site.z_coordinate()
    if z is not None:
        assert isinstance(z, int)


def test_site_is_zone(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site is_zone is a boolean."""
    _device, site = device_and_site
    iz = site.is_zone()
    if iz is not None:
        assert isinstance(iz, bool)


def test_site_x_extent(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site x extent is a positive integer."""
    _device, site = device_and_site
    xe = site.x_extent()
    if xe is not None:
        assert isinstance(xe, int)
        assert xe > 0


def test_site_y_extent(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site y extent is a positive integer."""
    _device, site = device_and_site
    ye = site.y_extent()
    if ye is not None:
        assert isinstance(ye, int)
        assert ye > 0


def test_site_z_extent(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site z extent is a positive integer."""
    _device, site = device_and_site
    ze = site.z_extent()
    if ze is not None:
        assert isinstance(ze, int)
        assert ze > 0


def test_site_module_index(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site module index is a non-negative integer."""
    _device, site = device_and_site
    mi = site.module_index()
    if mi is not None:
        assert isinstance(mi, int)
        assert mi >= 0


def test_site_submodule_index(device_and_site: tuple[Device, Site]) -> None:
    """Test that the site submodule index is a non-negative integer."""
    _device, site = device_and_site
    smi = site.submodule_index()
    if smi is not None:
        assert isinstance(smi, int)
        assert smi >= 0


def test_operation_name(device_and_operation: tuple[Device, Operation]) -> None:
    """Test that the operation name is a non-empty string."""
    _device, operation = device_and_operation
    name = operation.name()
    assert isinstance(name, str)
    assert len(name) > 0


def test_operation_qubits_num(device_and_operation: tuple[Device, Operation]) -> None:
    """Test that the operation qubits number is a positive integer."""
    _device, operation = device_and_operation
    qn = operation.qubits_num()
    if qn is not None:
        assert isinstance(qn, int)
        assert qn > 0


def test_operation_parameters_num(device_and_operation: tuple[Device, Operation]) -> None:
    """Test that the operation parameters number is a non-negative integer."""
    _device, operation = device_and_operation
    on = operation.parameters_num()
    if on is not None:
        assert isinstance(on, int)
        assert on >= 0


def test_operation_duration(device_and_operation: tuple[Device, Operation]) -> None:
    """Test that the operation duration is a positive integer."""
    _device, operation = device_and_operation
    dur = operation.duration()
    if dur is not None:
        assert isinstance(dur, int)
        assert dur > 0


def test_operation_fidelity(device_and_operation: tuple[Device, Operation]) -> None:
    """Test that the operation fidelity is a float between 0 and 1."""
    _device, operation = device_and_operation
    fid = operation.fidelity()
    if fid is not None:
        assert isinstance(fid, float)
        assert 0.0 <= fid <= 1.0


def test_operation_interaction_radius(device_and_operation: tuple[Device, Operation]) -> None:
    """Test that the operation interaction radius is a non-negative integer."""
    _device, operation = device_and_operation
    ir = operation.interaction_radius()
    if ir is not None:
        assert isinstance(ir, int)
        assert ir >= 0


def test_operation_blocking_radius(device_and_operation: tuple[Device, Operation]) -> None:
    """Test that the operation blocking radius is a non-negative integer."""
    _device, operation = device_and_operation
    br = operation.blocking_radius()
    if br is not None:
        assert isinstance(br, int)
        assert br >= 0


def test_operation_idling_fidelity(device_and_operation: tuple[Device, Operation]) -> None:
    """Test that the operation idling fidelity is a float between 0 and 1."""
    _device, operation = device_and_operation
    idf = operation.idling_fidelity()
    if idf is not None:
        assert isinstance(idf, float)
        assert 0.0 <= idf <= 1.0


def test_operation_is_zoned(device_and_operation: tuple[Device, Operation]) -> None:
    """Test that the operation is_zoned is a boolean."""
    _device, operation = device_and_operation
    iz = operation.is_zoned()
    if iz is not None:
        assert isinstance(iz, bool)


def test_operation_sites(device_and_operation: tuple[Device, Operation]) -> None:
    """Test that the operation sites is a non-empty list of Site objects."""
    device, operation = device_and_operation
    sites = operation.sites()
    if sites is not None:
        assert isinstance(sites, list)
        assert len(sites) > 0
        assert all(isinstance(site, Site) for site in sites)
        device_sites = device.sites()
        assert all(site in device_sites for site in sites)


def test_operation_mean_shuttling_speed(device_and_operation: tuple[Device, Operation]) -> None:
    """Test that the operation mean shuttling speed is a positive integer."""
    _device, operation = device_and_operation
    mss = operation.mean_shuttling_speed()
    if mss is not None:
        assert isinstance(mss, int)
        assert mss > 0
